File: biosec/application/ledger.java
import java.util.ArrayList;
import java.util.List;

public class Ledger {

    public static class Transaction {
        private int accountId;
        private String type; // "DEPOSIT" or "WITHDRAW"
        private double amount;

        public Transaction(int accountId, String type, double amount) {
            this.accountId = accountId;
            this.type = type;
            this.amount = amount;
        }

        public int getAccountId() {
            return accountId;
        }

        public String getType() {
            return type;
        }

        public double getAmount() {
            return amount;
        }

        public void display() {
            System.out.printf("AccountID: %d, Type: %s, Amount: %.2f\n", accountId, type, amount);
        }
    }

    private List<Transaction> transactions;

    public Ledger() {
        transactions = new ArrayList<>();
    }

    // เพิ่ม Transaction ใหม่
    public void addTransaction(int accountId, String type, double amount) {
        Transaction tx = new Transaction(accountId, type, amount);
        transactions.add(tx);
        System.out.println("Transaction added successfully.");
    }

    // แสดง Transaction ทั้งหมด
    public void showLedger() {
        if (transactions.isEmpty()) {
            System.out.println("Ledger is empty.");
            return;
        }
        System.out.println("\n--- Ledger ---");
        for (Transaction tx : transactions) {
            tx.display();
        }
    }

    // ดึง Transaction ทั้งหมด (สำหรับ Database integration)
    public List<Transaction> getTransactions() {
        return transactions;
    }
}

File: biosec/program/ledger.c
#include <stdio.h>
#include <string.h>
#include "ledger.h"

#define MAX_TRANSACTIONS 1000
#define LEDGER_FILE "ledger.dat"

typedef struct {
    int account_id;
    char type[20]; // "DEPOSIT" / "WITHDRAW"
    double amount;
} Transaction;

static Transaction transactions[MAX_TRANSACTIONS];
static int transaction_count = 0;

// โหลด ledger จากไฟล์
void initialize_ledger() {
    FILE *file = fopen(LEDGER_FILE, "rb");
    if (file) {
        fread(&transaction_count, sizeof(int), 1, file);
        fread(transactions, sizeof(Transaction), transaction_count, file);
        fclose(file);
    } else {
        transaction_count = 0;
    }
}

// บันทึก ledger ลงไฟล์
void save_ledger() {
    FILE *file = fopen(LEDGER_FILE, "wb");
    if (file) {
        fwrite(&transaction_count, sizeof(int), 1, file);
        fwrite(transactions, sizeof(Transaction), transaction_count, file);
        fclose(file);
    }
}

// เพิ่ม transaction ใหม่
void add_transaction(int account_id, const char *type, double amount) {
    if (transaction_count >= MAX_TRANSACTIONS) {
        printf("Ledger full! Cannot add more transactions.\n");
        return;
    }
    transactions[transaction_count].account_id = account_id;
    strncpy(transactions[transaction_count].type, type, sizeof(transactions[transaction_count].type) - 1);
    transactions[transaction_count].type[sizeof(transactions[transaction_count].type) - 1] = '\0';
    transactions[transaction_count].amount = amount;
    transaction_count++;
    save_ledger(); // save every transaction
}

// แสดง ledger ทั้งหมด
void show_ledger() {
    if (transaction_count == 0) {
        printf("Ledger is empty.\n");
        return;
    }

    printf("\n%-10s %-10s %-10s\n", "AccountID", "Type", "Amount");
    printf("-------------------------------\n");
    for (int i = 0; i < transaction_count; i++) {
        printf("%-10d %-10s %.2f\n", 
            transactions[i].account_id, 
            transactions[i].type, 
            transactions[i].amount);
    }
}

